### PR TITLE
Add Keycloak documentation

### PR DESCRIPTION
**Pull request by nebilbenmabrouk**
Add documentation for ProActive access control against Keycloak

### CODE REVIEW

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 1]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+In order to enact ProActive access control against a Keycloak server, two steps must be accomplished: (i) deploying and configuring a Keycloak server and (ii) configuring ProActive to connect to the Keycloak server and perform access control against it.
+
+==== Configuring Keycloak server
+
+WARNING: The configuration steps mentioned below provide an example of the basic steps needed to configure ProActive access control against Keycloak.
+For advanced configurations, please refer to the https://www.keycloak.org/documentation[Keycloak documentation].
+
+===== Pre-requisites
+1. A Keycloak server must be up and in operational mode so ProActive can interact with it.
+2. The administrator of the Keycloak should select a security realm (or create a new one if needed).
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2333]: The Keycloak server's administrator should

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 2]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+
+===== Configuring a new Keycloak client for ProActive
+The administrator of the Keycloak have to configure a new client for ProActive in the chosen realm.
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2336]: The Keycloak server's administrator have to

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 3]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+A client is an external software application to be protected using Keycloak.
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2337]: to be protected using Keycloak => not very clear.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 4]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+* The property `Name` specifies display name of the client  (e.g., ProActive Workflows and Scheduling).
+* The property `Description` specifies description of the client (e.g., ProActive workflows engine).
+
+.General settings of Keycloak client
+image::KeycloakClientGeneralConfig.png[Keycloak Client General Configuration,600,400]
+
+===== Configuring the authentication mechanism of the client
+In the section *Capability config* of the client configuration, the parameters following must be properly set (as shown in the screenshot below):
+
+* The property `Client authentication` must be enabled (i.e., set to On).
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2351]: set to `On`.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 5]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+
+.Authentication settings of the Keycloak client
+image::KeycloakClientAuthConfig.png[Keycloak Client Authentication Configuration,600,400]
+
+[[_keycloak_client_roles]]
+===== Configuring the client roles
+In the *Roles* tab of the client configuration, click on the button *Create role* and add the following parameters for each user role needed in ProActive.
+
+* The property `Role name` specifies name of the role. It will be used later on when <<_keycloak_proactive_privileges, configuring ProActive privileges>>.
+* The property `Composite` should be set to false.
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2363]: set to `false`.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 6]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+In order to enable ProActive access control against a Keycloak server, the following configuration steps must be accomplished at ProActive side:
+
+===== Configuring the authentication method for ProActive scheduler and Resource Manager
+The Keycloak authentication method must be selected for the scheduler and resource manager, as described in <<_select_authentication_method,Select authentication method>>.
+
+===== Configuring ProActive properties for Keycloak
+The configuration file config/authentication/keycloak.cfg must be edited in order the set the proper values of the following properties:
+
+* The property `keycloak.url` should be set to the URL of the Keycloak server (already configured for ProActive).
+* The property `keycloak.https.allow_any_certificate` must be set to true when the keycloak server operates in TLS mode and uses a non-trusted (e.g., self-signed) TLS/SSL certificate. Otherwise, this property is set to false.
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2390]: set to `true`

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 7]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+* The property `keycloak.use_resource_role_mappings` is set to true when the user roles are determined based on the information block `resource_access` in Keycloak access token. When it is set to false, the user roles are determined based on the information block `realm_access` in Keycloak access token.
+* Using the fall back to file authentication: ProActive can be configured to use simultaneously file-based authentication and Keycloak-based access control. In this scenario, ProActive Scheduler checks at first user credentials (login/password) and group memberships in
+the login and group files, as performed in the <<_file,File login method>>. If the user or group attribution is not found in the login/group files, the user credentials and roles will be checked via Keycloak. There are three configuration options to control the fallback mechanism:
+** `pa.keycloak.authentication.fallback`: if set to `true`, the user credentials will be checked first in the login file, and then in the Keycloak server. If set to `false`, only the Keycloak server will be queried.
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2400]: this should be set to true by default due to internal accounts.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 8]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+** `pa.keycloak.group.membership.fallback`: if set to `true`, group memberships will be read from the Keycloak server if the user exists, and then from the group file. In that sense, a user will be affected to groups defined in both file and Keycloak server. If set to `false`, only the Keycloak server will be queried.
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2401]: should also be set to true by default

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 9]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+
+# Authentication fallback Properties
+# check user password and group membership in login and group files
+# (as performed in FileLogin method), if user is not found in Keycloak.
+#
+# It use pa.[scheduler.core|RM].defaultloginfilename to authenticate user and
+# pa.[scheduler.core|RM].defaultgroupfilename files to check user membership.
+# In any case, no fallback is performed if user is found in Keycloak but has entered an incorrect password).
+# true or false.
+pa.keycloak.authentication.fallback=false
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2459]: should at least be set to true in the example configuration

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 10]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+
+# check user group membership in group file (as performed in FileLogin method)
+# if user is not found in its corresponding requested Keycloak group.
+# true or false.
+pa.keycloak.group.membership.fallback=false
```
Comment (fviale) [inline src/docs/admin/ProActiveAdminGuide.adoc:2464]: should at least be set to true in the example configuration

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 2 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 3 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Changed to 
`A client is an external software application for which the access control is managed via Keycloak.
`

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 5 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 6 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 7 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: I mentioned: If set to `true` (which is the default value)

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 8 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: I mentioned: 
If set to `true` (which is the default value)

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 9 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 10 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 11]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+[[_keycloak]]
+=== Keycloak-based access control
+https://www.keycloak.org/[Keycloak] is an access control solution that allows single sign-on with identity and access management aimed at modern applications and services.
+ProActive is able to connect to an existing Keycloak server, and perform user authentication and authorization against it.
+
+ProActive uses the OpenID Connect (OIDC) protocol to communicate with Keycloak and acquire an access tokens for a given user.
+It further parses the obtained access token and extracts from it the user information, including the user role.
+Based on the user role, ProActive determines the set of privileges and authorizations to be granted to the user.
+
+In order to enact ProActive access control against a Keycloak server, two steps must be accomplished: (i) deploying and configuring a Keycloak server and (ii) configuring ProActive to connect to the Keycloak server and perform access control against it.
```
Comment (AlexFalessi) [inline src/docs/admin/ProActiveAdminGuide.adoc:2324]: In order to perform user authentication against a Keycloak server...
(To reuse you terms from above)

**Review by fviale — APPROVED** [review]

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 12]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+For advanced configurations, please refer to the https://www.keycloak.org/documentation[Keycloak documentation].
+
+===== Pre-requisites
+1. A Keycloak server must be up and in operational mode so ProActive can interact with it.
+2. The Keycloak server's administrator should select a security realm (or create a new one if needed).
+
+===== Configuring a new Keycloak client for ProActive
+The Keycloak server's administrator have to configure a new client for ProActive in the chosen realm.
+A client is an external software application for which the access control is managed via Keycloak.
+In the section *General Settings* og the configuration, the parameters following must be properly set (as shown in the screenshot below):
```
Comment (AlexFalessi) [inline src/docs/admin/ProActiveAdminGuide.adoc:2338]: In the *General Settings* section of the configuration

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 11 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 13]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+
+WARNING: The configuration steps mentioned below provide an example of the basic steps needed to configure ProActive access control against Keycloak.
+For advanced configurations, please refer to the https://www.keycloak.org/documentation[Keycloak documentation].
+
+===== Pre-requisites
+1. A Keycloak server must be up and in operational mode so ProActive can interact with it.
+2. The Keycloak server's administrator should select a security realm (or create a new one if needed).
+
+===== Configuring a new Keycloak client for ProActive
+The Keycloak server's administrator have to configure a new client for ProActive in the chosen realm.
```
[the remark this answers is not in the document]
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 14]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+A client is an external software application for which the access control is managed via Keycloak.
+In the section *General Settings* og the configuration, the parameters following must be properly set (as shown in the screenshot below):
+
+* The property `Client type` specifies the security protocol of the client application. It must be set to *OpenID Connect*.
+* The property `Client ID` specifies the ID of the client application (e.g., ProActive).
+* The property `Name` specifies display name of the client  (e.g., ProActive Workflows and Scheduling).
```
Comment (AlexFalessi) [inline src/docs/admin/ProActiveAdminGuide.adoc:2342]: The property `Name` specifies the displayed name ...

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 12 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc]: I prefer keeping the bold format as it more visible. Thanks Alex.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 15]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+* The property `Description` specifies description of the client (e.g., ProActive workflows engine).
```
Comment (AlexFalessi) [inline src/docs/admin/ProActiveAdminGuide.adoc:2343]: * The property `Description` specifies **the** description ...

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 14 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc:2342]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc (hunk 15 above)
Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc:2343]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 16]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+
+.General settings of Keycloak client
+image::KeycloakClientGeneralConfig.png[Keycloak Client General Configuration,600,400]
+
+===== Configuring the authentication mechanism of the client
+In the section *Capability config* of the client configuration, the parameters following must be properly set (as shown in the screenshot below):
```
Comment (AlexFalessi) [inline src/docs/admin/ProActiveAdminGuide.adoc:2349]: ..., the following parameters must be ...

Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc:2349]: Done.

File: src/docs/admin/ProActiveAdminGuide.adoc [hunk 17]
```diff
@@ -2308,6 +2312,163 @@ In case of PAM users, the login file should not be modified as password authenti
...
+* The property `keycloak.public.key` must have as value the content of public encryption key of Keycloak (e.g., public RSA key). This key is used to decrypt the access token returned by Keycloak.
+* The property `keycloak.principal_attribute` is the parameter in Keycloak access token based on which the user name is determined (e.g., `name` or `given_name`).
+* The property `keycloak.use_resource_role_mappings` is set to `true` when the user roles are determined based on the information block `resource_access` in Keycloak access token. When it is set to `false`, the user roles are determined based on the information block `realm_access` in Keycloak access token.
+* Using the fall back to file authentication: ProActive can be configured to use simultaneously file-based authentication and Keycloak-based access control. In this scenario, ProActive Scheduler checks at first user credentials (login/password) and group memberships in
```
Comment (AlexFalessi) [inline src/docs/admin/ProActiveAdminGuide.adoc:2398]: fall back -> fallback

Comment (nebilbenmabrouk) [reply src/docs/admin/ProActiveAdminGuide.adoc:2398]: Done.

**Review by AlexFalessi — APPROVED** [review]
Small comments. Approved

**Pull request merged**